Common/Thread/ParallelLoop.cpp

```diff
@@ -0,0 +1,142 @@
+#include <cstring>
+
+#include "Common/Thread/ParallelLoop.h"
+#include "Common/CPUDetect.h"
+
+class LoopRangeTask : public Task {
+public:
+	LoopRangeTask(WaitableCounter *counter, const std::function<void(int, int)> &loop, int lower, int upper)
+		: counter_(counter), loop_(loop), lower_(lower), upper_(upper) {}
+
+	void Run() override {
+		loop_(lower_, upper_);
+		counter_->Count();
+	}
+
+	std::function<void(int, int)> loop_;
+	WaitableCounter *counter_;
+
+	int lower_;
+	int upper_;
+};
+
+WaitableCounter *ParallelRangeLoopWaitable(ThreadManager *threadMan, const std::function<void(int, int)> &loop, int lower, int upper, int minSize) {
+	if (minSize == -1) {
+		minSize = 1;
+	}
+
+	int numTasks = threadMan->GetNumLooperThreads();
+
+	int range = upper - lower;
+	if (range <= 0) {
+		// Bad range. A finished counter allocated.
+		return new WaitableCounter(0);
+	}
+
+	if (range <= numTasks) {
+		// Just assign one task per thread, as many as we have.
```

Comment:
Hm, I think we should still take `minSize` into account.  I would say:

```c++
if (range <= minSize) {
   // do on current thread only?
} else if (range <= numTasks) {
  // we can probably put on separate threads, might be nice to keep it lumpy (we don't know if the threads are actually free.)
} else {
```

Maybe this doesn't matter much in practice because the callers seem to check minSize.

-[Unknown]

Reply:
I think we can actually just fall through, the main logic should handle it appropriately .. will test.